### ISSUE TITLE
[HOLD for payment 2024-02-26][$500] [MEDIUM] Erroring thumbnail images try to reload infinitely

### ISSUE THREAD

**Issue by lindboe** [Expensify/App#34601]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.25-4
**Reproducible in staging?:** Yes
**Reproducible in production?:** It's worse on production, the app fails to load completely once I sign in.
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):** lizzi@infinitered.com
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** @lindboe
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1705427206739379?thread_ts=1705360862.915459&cid=C01GTK53T8Q

## Action Performed:

1. Have a request with a receipt image that produces a 404 when requested. (I am not sure how to reproduce this part outside of development, where you can give a thumbnail component an intentionally wrong URL.)
2. Open the app in a web browser and open the devtools. Observe that the requests for the image that can't be loaded are continuously made.

## Expected Result:

After the request returns a 404, the requests should stop being made.

## Actual Result:

Requests for the image are infinitely repeated.

## Workaround:

If this issue is encountered in production in its current state, the app can't be used at all, although I believe that particular issue may have been fixed by @cead22.

If this issue is encountered in staging, I don't think this prevents usage.

In development, it causes a connection/file descriptor leak as webpack generates more and more connections it never clears. This eventually leads to the developer's computer encountering errors and not working properly, including affecting internet access. (This must also be caused by a second issue with the app not clearing closed connections correctly).

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

In development:

https://github.com/Expensify/App/assets/5252268/701598d7-ee5c-45d6-9da8-4b60274683ad

Production screenshot:
<img width="2560" alt="Screenshot 2024-01-16 at 10 57 26 AM" src="https://github.com/Expensify/App/assets/5252268/f12829ca-a4e7-4189-8a7f-0033225a8d17">

## Root cause and considerations for proposals:

1. `Image`, which handles the load request, is memoized based on whether the identity of the `source` prop stays the same
2. `ImageWithSizeCalculation` is reconstructing `source` each render because it's creating the object in the prop. Pretty easy to fix with a `useMemo` instead:

```
    const source = useMemo(() => {
        return {uri: url};
    }, [url]);
```

This does NOT occur for the `Image/index.native.js` implementation.

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~010386784e89b6acc5</li>
        <li>Upwork Job ID: 1749005349524606976</li>
        <li>Last Price Increase: 2024-02-04</li>
<li>Automatic offers: </li>
<ul>
<li>paultsimura | Contributor | 28145121</li>
</ul></li>
    </ul>
</details>

**Comment by lindboe** [comment]
Update: I've started to be able to load data on production again, now behavior matches staging/dev.

**Comment by jliexpensify** [comment]
Hi @lindboe - cool, so all good here? Can I close this?

**Comment by lindboe** [comment]
@jliexpensify I'm not sure I understand the question? The original bug still exists, if you're referring to my update, note that in the original ticket production was breaking _worse_ than the other environments for me, in that it refused to load at all, but now all environments have the issue of constantly attempting to reload the 404ing image.

**Comment by jliexpensify** [comment]
Sorry @lindboe I misread your comment as the issue was fixed. I'll re-review this and see if I can reproduce.

**Comment by jliexpensify** [comment]
@tranvantoan-qn responded and can reproduce the `404 request` issue. Do you want to make a proposal in this GH?

**Comment by tranvantoan-qn** [comment]
@jliexpensify Sorry, I'm not a developer (I used to be a test contributor when the program was still open to testers). However, I've experienced that trying to reproduce issues can sometimes be impossible due to different conditions, so using some support tools can be very helpful. If you want to know more about how to use these tools to mimic scenarios like this one, I'm happy to help.

Once we can reproduce it either in a normal way or by using the tool, I'm sure that we can fix it and verify the soluton

**Comment by paultsimura** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
When an image cannot be loaded, requests are constantly being made.

### What is the root cause of that problem?
The root cause is in the `ImageWithSizeCalculation` component, which builds the `source` inside the render body:

https://github.com/Expensify/App/blob/15338b9c07c37a06b49247df79c1dc405e0beee8/src/components/ImageWithSizeCalculation.tsx#L76

So the component is re-rendered each time the request fails, and the request is sent again. This creates a loop.

### What changes do you think we should make in order to solve the problem?
Instead of building the source in-place, we should use the useMemo hook:

```js
const source = useMemo(() => ({uri: url}), [url]);
```

And use this `source` for `Image.source`.


### What alternative solutions did you explore? (Optional)

**Comment by jliexpensify** [comment]
Got it, thanks for the context @tranvantoan-qn! My biggest question is to understand how these `404` errors are generated and if they can be generated by actual users. The reason I ask is that we have pivoted to now focusing on project-specific bugs that impact performance/the UI so if it's going to affect actual users, we'd want to get it fixed. It's just unusual because I don't think anyone on our team, who use the app in real life, have experienced this with our requests.

**Comment by paultsimura** [comment]
@jliexpensify I have described the easiest way to reproduce the issue at the moment: 

> This can be easily (for now) reproduced when modifying the Distance request offline. The Distance receipt will be fixed in the https://github.com/Expensify/App/issues/34287, but for now, it allows us to test this endless requests loop.

https://github.com/Expensify/App/issues/34601#issuecomment-1902764996

**Comment by jliexpensify** [comment]
It looks like your proposal got posted as I was writing mine - it's great that we have a way to easily repro this one! If it's associated with Distance requests atm, we can fold this into #wave5.

**Comment by jliexpensify** [comment]
@parasharrajat just so I understand, is this the blank space you're referring to?

![image](https://github.com/Expensify/App/assets/30815269/1ed0d2e7-26b6-467b-902d-0f3387921489)

With this statement:

>This should not happen, image should always load.

Are you proposing that we won't show anything blank (image-wise), but instead just go to automatically loading the image?

**Comment by paultsimura** [comment]
> If image loading timeouts, we should show something instead of these blanks.

Yes, that's what I meant in the [alternative solution](https://github.com/Expensify/App/issues/34601#issuecomment-1902764996): to have some kind of a fallback image. In my case, I used the `ReceiptGeneric`, but it can be something more generic indicating that the image failed to load.

**Comment by jliexpensify** [comment]
@dubielzyk-expensify @shawnborton @dannymcclain 

![image](https://github.com/Expensify/App/assets/30815269/e877c562-5416-4d9b-824d-1426e3ba04de)

TL;DR - we're working on this bug around failed receipt load images. [We currently show blank images](https://github.com/Expensify/App/issues/34601#issuecomment-1913763923) with these timeouts, but we're wondering if we should instead have a fallback image? Any ideas/suggestions?

Thanks!

**Comment by dannymcclain** [comment]
Yes we probably should! I think we should probably have generic fallback images for offline and general-we-didn't-load-it. I can generate some ideas probably tomorrow or another designer can run with it! There are some offline placeholder ideas [in this Figma file](https://www.figma.com/file/WlrZyRCTQ0LFoKcjSmtTsd/General-UI-Patterns?type=design&node-id=2686-34760&mode=design&t=AVBg6Vw3KPwu8uAN-11) if we want to use those as starting points. But note that those do not use our product colors (which they should) _and_ I think we need less-specific images. We should just have basically one that handles offline and one that handles "didn't load" for all the things. (In my opinion anyways, feel free to push back on that lol.)

**Comment by dubielzyk-expensify** [comment]
I really like where you landed on this Danny so I'm keen to implement it. I felt this was the best of many great ones:
![CleanShot 2024-01-30 at 08 52 11@2x](https://github.com/Expensify/App/assets/145416827/f7aaed79-cdf7-4998-a7c6-1eeeaaed6575)

Either way we can always change it down the line. I added a variation on the left (and tweaked to new colors) if we needed something more "image isn't loading" for generic use rather than just receipt image:
![CleanShot 2024-01-30 at 09 00 59@2x](https://github.com/Expensify/App/assets/145416827/221d043b-1172-499a-9f53-f85db582a964)

I still think I prefer what Danny made

**Comment by dannymcclain** [comment]
@shawnborton @dubielzyk-expensify

I think I'm going to propose a potentially spicy take 🌶️ 

- We use a single `offline` icon for all placeholders that didn't load because you're offline.
- We use generic `type` icons for all placeholders that didn't load for _whatever other reason_.

This keeps things super simple, no custom icon creations needed, and personally I feel like this still accomplishes the goal of letting users know "Something is supposed to be here, but it isn't." (We could honestly even _just_ use the "type" icons for all situations—offline or not. I don't think these need to be super duper literal. We just need to communicate to users that "There is a thing here, but it didn't load at this time.")

Super keen for y'all's thoughts! 🌶️ 🌭 😈 

![image](https://github.com/Expensify/App/assets/7516624/507678c5-066f-4266-82cc-b240dd704bcc)

**Comment by dubielzyk-expensify** [comment]
Love the simplicity of that! 👍

**Comment by paultsimura** [comment]
Thanks @parasharrajat, updated the [proposal](https://github.com/Expensify/App/issues/34601#issuecomment-1902764996).

**Comment by dannymcclain** [comment]
Seems like we're all down for [this treatment](https://github.com/Expensify/App/issues/34601#issuecomment-1917172519). Let me know if you all need any assets for this (file type, dimensions, etc.) and I can get them exported/uploaded for you.

I will note that those images are all using icons from our standard set of icons and using our standard product color variables, so if we can create them just using SVGs + code, I _feel_ like that would be best since we'll get automatic theme switching, more flexibility, and won't have to deal with static image files, BUT I'll defer to you more technical folks to let me know if that will/won't work.

**Comment by paultsimura** [comment]
> So we don't want to show a fallback image while offline.

Was there any further internal discussion on this? It's a bit different from what was suggested here: https://github.com/Expensify/App/issues/34601#issuecomment-1917172519

**Comment by paultsimura** [comment]
> your proposal is showing the fallback when we go offline instead the requirement is that we should this fallback only when image fails to load while offline

In my understanding, this code from my proposal would do exactly this:

```js
const source = useMemo(() => {
  if (!hasFailedToLoadRef.current) {
    return {uri: url};
  }
  
  if (isOffline) {
    return OFFLINE_IMG;
  }

  return FALLBACK_BASED_ON_TYPE
}, [url, hasFailedToLoadRef, isOffline, type]);
```

First, we early return a valid URL if the image **didn't fail** to load.
Only after that - if we didn't return early - meaning the image failed to load — return a fallback: either offline one, or the type-based one.
Shouldn't this work as the requirements say?

**Comment by MonilBhavsar** [comment]
Proposal looks good 👍 
Regarding this comment  https://github.com/Expensify/App/issues/34601#issuecomment-1919182738
I think SVG works well and we could use styling to support dark and light themes cc @paultsimura

**Comment by paultsimura** [comment]
Thanks. 
What is the process of adding these SVGs to the App? @dannymcclain

**Comment by dannymcclain** [comment]
> What is the process of adding these SVGs to the App?

I can certainly just upload them here—unless @shawnborton you have a better idea? Should we try to just use the icon svgs that already exist in the app?

**Comment by dannymcclain** [comment]
Ok that sounds good to me!

@paultsimura someone more technical can probably guide you to those better than I can, but they are located in the `assets/images` directory:

**Offline**: `assets/images/offline-cloud.svg`
**Document**: `assets/images/document.svg`
**Receipt**: `assets/images/receipt.svg`
**Image**: I could not find this one for reason, if anyone knows where it may be, let us know! Attaching the SVG file in case it's not already in there for some reason: [Image.svg.zip](https://github.com/Expensify/App/files/14181805/Image.svg.zip)

**Comment by MonilBhavsar** [comment]
@paultsimura how's it going with the PR?

**Comment by paultsimura** [comment]
> @paultsimura how's it going with the PR?

This one is quite complicated – working on it now. I'll create a draft to keep the discussion with the Design team there.

**Comment by MonilBhavsar** [comment]
That will be great, thank you!

**Comment by allroundexperts** [comment]
Can someone please assign this to me since I reviewed the PR? Thanks!

**Comment by jliexpensify** [comment]
I assume we unassign @parasharrajat ?

**Comment by paultsimura** [comment]
Deployed to prod: https://github.com/Expensify/App/pull/36214#issuecomment-1945393766

**Comment by allroundexperts** [comment]
## Checklist
1. https://github.com/Expensify/App/pull/24159
2. https://github.com/Expensify/App/pull/24159/files#r1493106476
3. N/A
4. I think this was an edge case that is not easily reproducible without altering the generated HTML or the source code. As such, a regression test might not make complete sense here.

**Comment by jliexpensify** [comment]
[No checklist needed](https://github.com/Expensify/App/issues/34601#issuecomment-1949545338)

Payment Summary:

- Contributor: @paultsimura $500 (PAID)
- C+: @allroundexperts $500 (New.Dot)

**Comment by JmillsExpensify** [comment]
$500 approved for @allroundexperts based on summary above.
